### DISCUSSION TITLE
support for fluidsynth midi device or Roland MT-32

### DISCUSSION THREAD

**Discussion by zbyti** [paulscottrobson/neo6502-firmware#209] [Ideas]
we can provide soundfonts with game for fluidsynth, or we cen support one of popular midi device like Roland MT-32 - then everybody have same experience.

at first, we can do it for the emulator, when a hardware solution appears, there may be games/demos for it ;)

**Comment by paulscottrobson** [comment]
Could add multiple basic sound channels to the emulator that the hardware now simply ignores. I think we can be fairly sure any sound solution would involve say at least 3 beepers. Anything else probably should wait so we're not locked in to one design.

**Comment by zbyti** [comment]
I think more about: how to use emulator and UEXT with https://wiki.archlinux.org/title/FluidSynth

as far as I understand, the Neo6502 emulator should open a port that I could listen to and transmit data to the midi server.

**Reply by paulscottrobson** [reply]
I would think this needs to be optional ; it's quite a hefty dependency.

**Reply by jviskari** [reply]
Having MIDI out would be quite handy. Basically it requires one of the following.
- software USB midi host where USB2MIDI adapter (such as Roland UM2 or sound module could be attached)
- I2C to MIDI midi bridge to UEXT
- Serial to MIDI bridge.

I think Neo6502 or emulator would not have to support any specific midi device, but just output general MIDI data. In other words, if there's a virtual or HW midi port available the data is forwarded there. Event easier would be just to make midi data available via local socket for 3rd party players. 

MidiOut API from 65C02 point of view is quite nice.  The system just needs to send 3 bytes out transparently via UEXT or USB midi.

I suppose crude beeper midi would work by just using note on and note off commands:

8n, note, velocity
9n, note, velocity

n is channel.

**Comment by astralaster** [comment]
I think this fits here: https://github.com/rppicomidi/usb_midi_host . This should extend tinyusb host to also support MIDI USB devices. I did not test it yet, but I really want USB MIDI support anyway.

**Comment by zbyti** [comment]
it would be nice to share the game with emulator and the .sf2 / .sf3 (sundfont) file so that everyone have the same music experience.

and this can be done by adding a library for support, e.g. fluidsynth.

on real hardware you could simply load these soundfonts into the sound bank of the midi device.

---

I wrote one code playing music on the beeper and that's enough :D

**Comment by paulscottrobson** [comment]
I don't think anyone would argue the beeper is pretty terrible :) I can't see why we can't do Midi (can we just use the serial port ?) and some sort of sound device.  My gut tells me HDMI Audio is going to consume resources I'm loathe to give up, when an I2C add on should be pretty cheap.

**Reply by zbyti** [reply]
I can't wait for the time then Neo6502 do the great music but... unfortunately my skills don't allow me to help with this matter.

**Comment by paulscottrobson** [comment]
I experimented with a PWM generator which worked rather well ; it could play samples audibly. I don't think we'll get samples, because the memory isn't there, but it should be possible to have some multi channels with some waveforms, or even emulating a real chip.

**Comment by erlendnagel** [comment]
Would a USB soundcard work? Or are these essentially 'unsupported' by the current USB implementation (HID + storage only?)

**Comment by paulscottrobson** [comment]
Probably. USB everything else works I think. Apparently there is a USB soundcard standard but not all devices support it, and obviously we couldn't run x86 drivers. 

I'd quite like the option of using the original hardware though. I tried the PCM code @astralaster found and was surprised by how good the sound from it was. Good enough for basic gaming SFX at least.

The sound system does allow for multiple targets, so handling such would be fairly transparent providing the drivers aren't too complicated.
